### ISSUE TITLE
path binding rendering before router component renders causes error

### ISSUE THREAD

**Issue by Ixonal** [Profiscience/ko-component-router#157]
I have a setup that's effectively as follows (in the way of component structure)

```html
<outer-component>
  <header-component>
    <a data-bind="path: '/some/path'">navigation link</a>
  </header-component>
  <content-container-component>
    <ko-component-router></ko-component-router>
  </content-container-component>
</outer-component>
```

So if both "header-component" and "content-container-component" are synchronous, the path binding will throw an error, since no Router instance has been created yet. The place it fails is in the "resolveHref" function when it tries to reference "router.base" (router is undefined). Would it be possible to have it create an interim href value if there is no Router instance present and then update when a Router is actually added to the page?

This issue can, however, be covered up by making "header-component" asyncronous and "content-container-component" synchronous, since that forces the router to be created before the path binding is used. That doesn't seem like a particularly good solution to me, though.

**Comment by matt-hensley** [comment]
try out this async binding from v 3.x:
```javascript
ko.bindingHandlers['async'] = {
    init(el, valueAccessor, bindings, viewModel, bindingCtx) {
        ko.tasks.schedule(() => ko.applyBindingsToNode(el, valueAccessor(), bindingCtx))
        return { controlsDescendantBindings: true }
    }
}
```
there are some timing difficulties like this lurking still.

**Comment by caseyWebb** [comment]
to clarify...

you'd create the above binding and use it as such

```html
<a data-bind="async: { path: '/some/path' }">navigation link</a>
```

The binding already waits for one ko render cycle to complete before attaching itself so that immediately adjacent routers are available — which means it's curious that it doesn't work when both are synchronous. I believe the further nesting in `<content-container-component>` may be to blame.

**Comment by Ixonal** [comment]
@caseyWebb Right, but that still seems kinda off, since it's relying on another binding, and it's still possible that wouldn't fix this issue (if all components are async, I would think the situation could still arise, even with the async binding). Be that as it may, I do have an interim solution, in making the header async and the rest sync. 

And yes, the problem does seem to be that the router is nested within another component. Since the "header-component" is first in the list of DOM nodes, and the component is rendered synchronously, its template (including the path bindings) are rendered before the router is rendered. 

In thinking about it, would it be possible to always have the head router exist, but just not attached to anything until the first "ko-component-router" is rendered, then attach it to that? Though, there would probably still be some weirdness if child router components come into play...

**Comment by caseyWebb** [comment]
Having the root router predefined could cause some issues since the routes don't have to be defined globally, and they aren't observable. However, p sure it'd be fairly trivial to export a promise from `router.js` that resolves when the root is initialized and use that in `binding.js`.

**Comment by Ixonal** [comment]
I like it. That promise could be helpful elsewhere as well when trying to do fancy stuff with the router.

**Comment by Ixonal** [comment]
@caseyWebb Hope you're still getting notifications on this. The fix looks to be working for the most part (thanks for the quick support, btw), but I am having a weird niche case issue. Every time, when I first start up the app (using Visual Studio, starting an IIS Express server), the first load of the page doesn't update the elements using the path binding. I've traced this to the first update call in the Router constructor not resolving for some reason (so the onInit methods never run). Subsequent reloads of the page run as expected, though. This behavior is consistent among every browser I've tested on. It doesn't look like it's the Promise polyfill (Bluebird) I'm using, as I took it out of the build and let Chrome use its native promise implementation, and had the same issue. Any ideas what's up?

Edit: Never mind, it seems to have fixed itself. I haven't altered anything, so I'm not sure what changed to fix it, but so long as it works, I'm good.
